README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# stochvolTMB

<!-- badges: start -->

[![CRAN\_Status\_Badge](https://www.r-pkg.org/badges/version-last-release/stochvolTMB)](https://cran.r-project.org/package=stochvolTMB)
[![CRAN RStudio mirror
downloads](https://cranlogs.r-pkg.org/badges/grand-total/stochvolTMB?color=blue)](https://r-pkg.org/pkg/stochvolTMB)
[![R
buildstatus](https://github.com/JensWahl/stochvolTMB/workflows/R-CMD-check/badge.svg)](https://github.com/JensWahl/stochvolTMB/actions)
[![License:
GPLv3](https://img.shields.io/badge/License-GPLv3-blue.svg)](https://www.gnu.org/licenses/gpl-3.0)
[![Lifecycle:experimental](https://img.shields.io/badge/lifecycle-experimental-orange.svg)](https://www.tidyverse.org/lifecycle/#experimental)
<!-- badges: end -->

`stochvolTMB` is a package for fitting stochastic volatility (SV) models
to time series data. It is inspired by the package
[stochvol](https://github.com/gregorkastner/stochvol), but parameter
estimates are obtained through optimization and not MCMC, leading to
significant speed up. It is built on [Template Model
Builder](https://github.com/kaskr/adcomp) for fast and efficient
estimation. The latent volatility is integrated out of the likelihood
using the Laplace approximation and automatic differentiation (AD) is
used for accurate evaluation of derivatives.

Four distributions for the observational error are implemented:

  - **Gaussian** - The classic SV model with Gaussian noise
  - **t** - t-distributed noise for heavy tail returns
  - **Leverage** - Extending the **Gaussian** model by allowing observed
    returns to be correlated with the latent volatility
  - **Skew-Gaussian** - Skew-Gaussian distributed noise for asymmetric
    returns

## Installation

To install the current stable release from CRAN, use

``` r
install.packages("stochvolTMB")
```

To install the current development version, use

``` r
# install.packages("remotes")
remotes::install_github("JensWahl/stochvolTMB")
```

If you would also like to build and view the vignette locally, use

``` r
remotes::install_github("JensWahl/stochvolTMB", dependencies = TRUE, build_vignettes = TRUE)
```

## Example

The main function for estimating parameters is `estimate_parameters`:

``` r
library(stochvolTMB, warn.conflicts = FALSE)

# load s&p500 data from 2005 to 2018
data(spy)

# find the best model using AIC 
gaussian <- estimate_parameters(spy$log_return, model = "gaussian", silent = TRUE)
t_dist <- estimate_parameters(spy$log_return, model = "t", silent = TRUE)
skew_gaussian <- estimate_parameters(spy$log_return, model = "skew_gaussian", silent = TRUE)
leverage <- estimate_parameters(spy$log_return, model = "leverage", silent = TRUE)

# the leverage model stands out with an AIC far below the other models
AIC(gaussian, t_dist, skew_gaussian, leverage)
#>               df       AIC
#> gaussian       3 -23430.57
#> t_dist         4 -23451.69
#> skew_gaussian  4 -23440.87
#> leverage       4 -23608.85

# get parameter estimates with standard error
estimates <- summary(leverage)
head(estimates, 10)
#>       parameter     estimate    std_error     z_value       p_value        type
#>  1:     sigma_y  0.008338412 0.0004163314  20.0283029  3.121144e-89 transformed
#>  2:     sigma_h  0.273443559 0.0182641070  14.9716359  1.125191e-50 transformed
#>  3:         phi  0.967721215 0.0043681868 221.5384240  0.000000e+00 transformed
#>  4:         rho -0.748695259 0.0322487815 -23.2162340 3.121690e-119 transformed
#>  5: log_sigma_y -4.786882463 0.0499293427 -95.8731319  0.000000e+00       fixed
#>  6: log_sigma_h -1.296660043 0.0667929683 -19.4131220  5.978190e-84       fixed
#>  7:   logit_phi  4.110221202 0.1375467861  29.8823500 3.337032e-196       fixed
#>  8:   logit_rho -1.939958912 0.1467670249 -13.2179481  6.912403e-40       fixed
#>  9:           h -0.536254072 0.5182192669  -1.0348015  3.007616e-01      random
#> 10:           h -0.207811236 0.4245258952  -0.4895137  6.244781e-01      random

# plot estimated volatility with 95 % confidence interval
plot(leverage, include_ci = TRUE, dates = spy$date)
```

<img src="man/figures/README-example-1.png" width="100%" />

Given the estimated parameters we can simulate future volatility and
log-returns using `predict`.

``` r
# simulate future prices with or without parameter uncertainty
pred = predict(leverage, steps = 10)

# Calculate the mean, 2.5% and 97.5% quantiles from the simulations
pred_summary = summary(pred, quantiles = c(0.025, 0.975), predict_mean = TRUE)

print(pred_summary)
#> $y
#>     time quantile_0.025 quantile_0.975          mean
#>  1:    1    -0.03811074     0.03725858 -1.605944e-04
#>  2:    2    -0.03613254     0.03807427  3.905524e-04
#>  3:    3    -0.03689599     0.03697331 -1.170203e-04
#>  4:    4    -0.03578552     0.03655235  1.860492e-04
#>  5:    5    -0.03547748     0.03626548  1.165658e-04
#>  6:    6    -0.03661546     0.03555252 -1.898271e-04
#>  7:    7    -0.03495112     0.03523121  1.582960e-04
#>  8:    8    -0.03567302     0.03514370  3.110404e-05
#>  9:    9    -0.03521115     0.03470416 -3.077596e-05
#> 10:   10    -0.03391924     0.03442448 -3.941700e-05
#> 
#> $h
#>     time quantile_0.025 quantile_0.975     mean
#>  1:    1    0.411180931       2.509780 1.450119
#>  2:    2    0.255810550       2.557694 1.401863
#>  3:    3    0.104986137       2.586031 1.354629
#>  4:    4    0.004664263       2.619603 1.310351
#>  5:    5   -0.120600217       2.660227 1.268050
#>  6:    6   -0.209691395       2.670629 1.227245
#>  7:    7   -0.306100384       2.657083 1.189820
#>  8:    8   -0.392420025       2.711739 1.153733
#>  9:    9   -0.469962600       2.714449 1.115334
#> 10:   10   -0.537930702       2.720761 1.079353
#> 
#> $h_exp
#>     time quantile_0.025 quantile_0.975       mean
#>  1:    1    0.010082253     0.02956799 0.01790438
#>  2:    2    0.009515924     0.03051188 0.01757710
#>  3:    3    0.008796049     0.03044210 0.01725803
#>  4:    4    0.008247518     0.03108103 0.01699651
#>  5:    5    0.007821668     0.03170945 0.01673759
#>  6:    6    0.007420323     0.03200784 0.01647688
#>  7:    7    0.007167451     0.03213425 0.01628790
#>  8:    8    0.006796787     0.03246457 0.01610369
#>  9:    9    0.006584324     0.03277949 0.01578397
#> 10:   10    0.006352601     0.03281460 0.01566125

# plot predicted volatility with 0.025 and 0.975 quantiles
plot(leverage, include_ci = TRUE, forecast = 50, dates = spy$d) +
  ggplot2::xlim(c(spy[.N, date] - 150, spy[.N, date] + 50))
#> Warning: Removed 3419 row(s) containing missing values (geom_path).

#> Warning: Removed 3419 row(s) containing missing values (geom_path).

#> Warning: Removed 3419 row(s) containing missing values (geom_path).
```

<img src="man/figures/README-unnamed-chunk-5-1.png" width="100%" />

## Shiny app

By running `demo()` you start a shiny application where you can visually
inspect the effect of choosing different models and parameter
configurations

``` r
demo()
```

![](man/figures/shinyApp.png)


NEWS.md
# stochvolTMB 0.2.0

* Added `predict` function that makes it possible to simulate future volatility and log-returns. 
* `plot` has a new `forecast` argument that makes it possible to include forecasted volatility. 
* The degrees of freedom for the `t` model is estimated on log scale and shifted by 2 so that it is guaranteed to be 
  greater than 2. This is to ensure that the variance is finite. 
* Changed parameter names from `phi_logit` and `rho_logit` to `logit_phi` and `logit_rho` to be consistent with other 
  parameter names. 
* Updated README.
* Updated documentation. 

# stochvolTMB 0.1.2

* CRAN release 

R/optSV.R
#' Construct objective function with derivatives using \link[TMB]{MakeADFun}
#'@param data Vector of observations.
#'@param model String specifying distribution of error term in observational equation.
#'@param ... Additional arguments passed to \link[TMB]{MakeADFun}.
#'@return List of components returned from \link[TMB]{MakeADFun}.
#'@export
#'@keywords internal
get_nll <- function(data, model = "gaussian", ...) {

  # Starting values for parameters
  param <- list(log_sigma_y = 0,
                log_sigma_h = 0, 
                logit_phi = 3,
                log_df_minus_two = if (model == "t") 2 else numeric(0),
                alpha = if (model %in% c("skew_gaussian")) 0 else numeric(0),
                logit_rho = if (model %in% c("leverage")) 0 else numeric(0),
                h = rep(0, length(data)))
  
  data <- list(y = data,
               model = ifelse(model == "gaussian", 0,
                               ifelse(model == "t", 1, 
                                      ifelse(model == "skew_gaussian", 2,
                                             ifelse(model == "leverage", 3, 4)))))
  
  obj <- TMB::MakeADFun(data = data, parameters = param, random = "h", DLL = "stochvolTMB", ...)
  
  return(obj)
  
}

#' Estimate parameters for the stochastic volatility model
#' 
#' @description 
#' Estimate parameters of a stochastic volatility model with a latent 
#' log-volatility following an 
#' autoregressive process of order one with normally distributed noise. 
#' The following distributions are implemented for the observed process: 
#' 
#' \itemize{
#' \item Gaussian distribution
#' \item t-distribution
#' \item Leverage: Gaussian distribution with leverage where the noise of the latent process is correlated with the
#'  observational distribution
#' \item Skew gaussian distribution
#' }
#' 
#' The parameters is estimated by minimizing the negative log-likelihood (nll) and the latent log-volatility is 
#' integrated out by applying the Laplace approximation.
#' 
#' @param data A vector of observations.
#' @param model A character specifying the model. Must be one of the following:
#'  "gaussian", "t", "leverage", "skew_gaussian".
#' @param opt.control An optional list of parameters for nlminb.
#' @param ... additional arguments passed to \link[TMB]{MakeADFun}.
#' 
#' @return Object of class \code{stochvolTMB}
#' 
#' @export 
#' 
#' @examples
#' \donttest{
#' # load data
#' data("spy")
#'
#' # estimate parameters 
#' opt <- estimate_parameters(spy$log_return, model = "gaussian")
#'
#' # get parameter estimates with standard error
#' estimates <- summary(opt)
#'
#' # plot estimated volatility with 95 % confidence interval
#' plot(opt, include_ci = TRUE)
#' }

estimate_parameters <- function(data, model = "gaussian", opt.control = NULL, ...) {
  
  if (!is.vector(data)) stop("data needs to be a vector")
  if (!is.character(model)) stop("model has to be a character")
  if (!(model %in% c("gaussian", "skew_gaussian", "t", "leverage"))) stop("This model is not implemented")
  
  # create TMB object
  obj <- get_nll(data, model, ...)
  
  # Optimize nll 
  fit <- stats::nlminb(obj$par, obj$fn, obj$gr, opt.control)
  
  # Calculate standard error for all parameters (including latent)
  rep <- TMB::sdreport(obj)
  opt <- list()
  class(opt) <- "stochvolTMB"

  opt$rep <- rep
  opt$obj <- obj
  opt$fit <- fit
  opt$nobs <- length(data)
  opt$model <- model

  return(opt)
}
  

#' @export
logLik.stochvolTMB <- function(object, ...) {
  val <- -object$fit$objective
  attr(val, "nobs") <- object$nobs
  attr(val, "df") <- length(object$fit$par)
  class(val) <- "logLik"
  val
}


#' Summary tables of model parameters 
#' 
#' Extract parameters, transformed parameters and latent log volatility along with standard error, z-value and p-value 
#' 
#' @rdname summary
#' @param object A \code{stochvolTMB} object.
#' @param ... Currently not used. 
#' @param report Parameters to report with uncertainty estimates. Can be any subset of "fixed", "transformed" or 
#' "random" (see \link[TMB]{summary.sdreport}). "fixed" 
#' report the parameters on the scale they were estimated, for example are all standard deviations estimated on 
#' log scale. "transformed" report all transformed parameters, for example estimated standard deviations 
#' transformed from log scale by taking the exponential. Lastly, "random"
#' report the estimated latent log-volatility. 
#' @return \code{data.table} with parameter estimates, standard error, z-value and approximated p-value.
#' @export 

summary.stochvolTMB <- function(object, ..., report = c("all", "fixed", "transformed", "random")) {
  
  # check argument
  report <- match.arg(report, several.ok = TRUE)
  
  if ("all" %in% report) report <- c("fixed", "transformed", "random")
  
  rep <- object$rep
  
  srep_fixed <- srep_report <- srep_random <- NULL
  column_names <- c("estimate", "std_error", "z_value", "p_value")
  
  if ("fixed" %in% report) {
    
    srep_fixed <- as.data.table(TMB::summary.sdreport(rep, select = c("fixed"), p.value = TRUE),
                                keep.rownames = "parameter")
    
    setnames(srep_fixed, 2:5, column_names)
    srep_fixed[, type := "fixed"]

  }
  
  if ("transformed" %in% report) {
    
    srep_report <- as.data.table(TMB::summary.sdreport(rep, select = c("report"), p.value = TRUE),
                                 keep.rownames = "parameter")
    
    setnames(srep_report, 2:5, column_names)
    srep_report[, type := "transformed"]

  }
  
  if ("random" %in% report) {
    srep_random <- as.data.table(TMB::summary.sdreport(rep, select = c("random"), p.value = TRUE),
                                 keep.rownames = "parameter")
    
    setnames(srep_random, 2:5, column_names)
    srep_random[, type := "random"]
  }
  
  
  srep <- rbindlist(list(srep_report, srep_fixed, srep_random))
  
  if (nrow(srep) == 0) {
    warning("no or empty summary selected via 'report = %s'", 
            deparse(report))
  }
  
  return(srep)
  
  
}




#' @export
print.stochvolTMB <- function(x, ...) {
  rep <- summary(x, report = c("fixed", "transformed"))
  
  cat("\nMaximum likelihood estimates for ", x$model, " model based on ", x$nobs, " observations.\n\n", sep = "")
  cat("Parameters:\n")
  cat("standard deviation of latent variable           sigma_h = ", rep[parameter == "sigma_h", estimate], "\n", sep = "")
  cat("persistence of latent variable                  phi = ", rep[parameter == "phi", estimate], "\n", sep = "")
  cat("standard deviation of observed variable         sigma_y = ", rep[parameter == "sigma_y", estimate], "\n", sep = "")
  if ("df" %in% rep[, parameter])
    cat("degrees of freedom of observed variable        df =", rep[parameter == "df", estimate], "\n")
  if ("rho" %in% rep[, parameter]) 
    cat("leverage effect                                 rho =", rep[parameter == "rho", estimate], "\n")
  if ("alpha" %in% rep[, parameter]) 
    cat("skewness of observed variable                   alpha = ", rep[parameter == "alpha", estimate], "\n", sep = "")
  
  return(invisible(x))
}


#' Predict future returns and future volatilities
#' 
#' Takes a \code{stochvolTMB} object and produces draws from the predictive distribution of the latent volatility 
#' and future log-returns. 
#' 
#' @param object A \code{stochvolTMB} object returned from \code{\link{estimate_parameters}}.
#' @param steps Integer specifying number of steps to predict. 
#' @param nsim Number of draws from the predictive distribution.
#' @param include_parameters  Logical value indicating if fixed parameters 
#' should be simulated from their asymptotic distribution, i.e. 
#' multivariate normal with inverse hessian as covariance matrix. 
#' @param ... Not is use. 
#' @return List of simulated values from the predictive distribution of the latent volatilities and log-returns.
#' @export

predict.stochvolTMB <- function(object, steps = 1L, nsim = 10000, include_parameters = TRUE, ...) {
  
  time <- std_error <- NULL
  
  
  if (!inherits(object, "stochvolTMB")) {
    stop("`object` has to be of class `stochvolTMB`")
  }
  
  if (steps < 1) {
    stop("`steps` has to be greater or equal to 1")
  }
  
  # We need an extra step for the volatility to predict the last log-return
  if (object$model == "leverage") steps <- steps + 1
  
  # Get parameter estimates 
  rep <- summary(object)

  # Simulate parameters
  if (include_parameters) {
    
    sim_parameters <- simulate_parameters(object, nsim = nsim)
    if (object$model != "leverage") rho <- 0 else rho <- sim_parameters[, which(colnames(sim_parameters) == "rho")]
    if (object$model != "t") df <- Inf else df <-  sim_parameters[, which(colnames(sim_parameters) == "df")]
    if (object$model != "skew_gaussian") alpha <- 0 else alpha <-  sim_parameters[, which(colnames(sim_parameters) == "alpha")]
    sigma_y <-  sim_parameters[, which(colnames(sim_parameters) == "sigma_y")]
    sigma_h <- sim_parameters[, which(colnames(sim_parameters) == "sigma_h")]
    phi <- sim_parameters[, which(colnames(sim_parameters) == "phi")]
    
  } else {
    
    if (object$model != "leverage") rho <- 0 else rho <- rep[parameter == "rho", estimate]
    if (object$model != "t") df <- Inf else df <- rep[parameter == "df", estimate]
    if (object$model != "skew_gaussian") alpha <- 0 else df <- rep[parameter == "alpha", estimate]
    sigma_y <- rep[parameter == "sigma_y", estimate]
    sigma_h <- rep[parameter == "sigma_h", estimate]
    phi <- rep[parameter == "phi", estimate]
    
  }
  
  # Get last estimated volatility and simulate from its distribution to get nsim starting points
  h_last <- rep[.N, list(estimate, std_error)]
  h_pred <- matrix(0, nrow = steps, ncol = nsim)
  y_pred <- matrix(0, nrow = steps, ncol = nsim)
  h_start <- stats::rnorm(nsim, h_last$estimate, h_last$std_error)
  
  # Predict volatility 
  for (i in 1:steps) {
    if (i == 1) {
      h_pred[i, ] <- stats::rnorm(nsim, phi * h_start, sigma_h)
    } else {
      h_pred[i, ] <- stats::rnorm(nsim, phi * h_pred[i - 1, ], sigma_h)
    }
  }  
  
  if (object$model == "leverage") steps <- steps - 1
  
  # Predict log-returns
  for (i in 1:steps) {
    
    if (object$model == "skew_gaussian") {
      
      delta <- alpha / sqrt(1 + alpha^2)
      omega <- 1 / sqrt(1 - 2 * delta^2 / pi)
      xi <- -omega * delta * sqrt(2 / pi)
      y_pred[i, ] <- exp(h_pred[i, ] / 2) * sigma_y * sn::rsn(n = nsim, alpha = alpha, xi = xi, omega = omega)
      
    } else if (object$model == "t") {
      
      y_pred[i, ] <- exp(h_pred[i, ] / 2) * sigma_y * sqrt((df - 2) / df) * stats::rt(nsim, df = df)      

    } else if (object$model == "leverage") {
      
      y_pred[i, ] <- sigma_y * exp(h_pred[i, ] / 2) * (rho / sigma_h * (h_pred[i + 1, ] - phi * h_pred[i, ]) + sqrt(1 - rho^2) * stats::rnorm(nsim))
      
    } else {
      
      y_pred[i, ] <- exp(h_pred[i, ] / 2) * sigma_y * stats::rnorm(nsim)
      
    }
  }
  
  if (object$model == "leverage") {
    h_pred <- h_pred[1:steps, , drop = FALSE] 
    y_pred <- y_pred[1:steps, , drop = FALSE]
  }
    
  
  h_exp <- sigma_y * exp(h_pred / 2) 

  res <- list(y = y_pred, h = h_pred, h_exp = h_exp)
  attr(res, "steps") <- steps
  attr(res, "nsim") <- nsim
  class(res) <- "stochvolTMB_predict"
  
  return(res)
}

#' Calculate quantiles based on predictions from the predictive distribution
#' @param object A \code{stochvolTMB_summary} object.
#' @param ... Not used. 
#' @param quantiles A numeric vector specifying which quantiles to calculate. 
#' @param predict_mean bool. Should the mean be predicted? 
#' @return A list of \code{data.table}s. One for \code{y}, \code{h} and \code{h_exp}. 
#' @export
summary.stochvolTMB_predict <- function(object, ..., quantiles = c(0.025, 0.975), predict_mean = TRUE) {
  
  time <- NULL
  
  if (!inherits(object, "stochvolTMB_predict")) {
    stop("`object` should be of class stochvolTMB_predict")
  }
  
  if (length(quantiles) < 1) {
    stop("`quantiles` has to be of length 1 or greater")
  }
  
  if (!all(quantiles <= 1 & quantiles >= 0)) {
    stop("All elements of `quantiles` have to be between 0 and 1")
  }
  
  # If quantile has length 1 it returns a vector, not a matrix. When cast to matrix it will be a one column matrix
  # and should not be transposed
  res_quant <- lapply(object, function(x) {
    
    if (length(quantiles) == 1) {
      data.table(as.matrix(apply(x, 1, stats::quantile, probs = quantiles)))
    } else {
      data.table(t(apply(x, 1, stats::quantile, probs = quantiles)))
    }
  })
  
  lapply(res_quant, setnames, new = paste0("quantile_", quantiles))
  lapply(res_quant, function(x) x[, time := 1:.N])
  lapply(res_quant, data.table::setcolorder, "time")
  
  
  if (predict_mean) {
    res_mean <- lapply(object, function(x) apply(x, 1, mean))
    lapply(1:length(res_quant), function(x) res_quant[[x]][, mean := res_mean[[x]]])
  }
  
  return(res_quant)
  
}
